pkg/apiserver/rest/usecase/rbac.go

```diff
@@ -0,0 +1,783 @@
+						pathName: "revision",
+					},
+					"envBinding": {
+						pathName: "envName",
+					},
+					"trigger": {},
+				},
+			},
+			"environment": {
+				pathName: "envName",
+			},
+			"workflow": {
+				pathName: "workflowName",
+			},
+			"role":                {},
+			"permPolicy":          {},
+			"projectUser":         {},
+			"applicationTemplate": {},
+		},
+		pathName: "projectName",
+	},
+	"cluster": {
+		pathName: "clusterName",
+		subResources: map[string]resourceMetadata{
+			"namespace": {},
+		},
+	},
+	"addon": {
+		pathName: "addonName",
+	},
+	"addonRegistry": {
+		pathName: "addonRegName",
+	},
+	"target": {
+		pathName: "targetName",
+	},
+	"user":          {},
+	"role":          {},
+	"permPolicy":    {},
+	"systemSetting": {},
+	"definition":    {},
+}
+
+var existResourcePaths = convert(ResourceMaps)
+
+type resourceMetadata struct {
+	subResources map[string]resourceMetadata
+	pathName     string
+}
+
+func checkResourcePath(resource string) (string, error) {
+	path := ""
+	exist := 0
+	lastResourceName := resource[strings.LastIndex(resource, "/")+1:]
+	for key, erp := range existResourcePaths {
+		allMatchIndex := strings.Index(key, fmt.Sprintf("/%s/", resource))
+		index := strings.Index(erp, fmt.Sprintf("/%s:", lastResourceName))
+		if index > -1 && allMatchIndex > -1 {
+			pre := erp[:index+len(lastResourceName)+2]
+			next := strings.Replace(erp, pre, "", 1)
+			nameIndex := strings.Index(next, "/")
+			if nameIndex > -1 {
+				pre += next[:nameIndex]
+			}
+			if pre != path {
+				exist++
+			}
+			path = pre
+		}
+	}
+	path = strings.TrimSuffix(path, "/")
+	path = strings.TrimPrefix(path, "/")
+	if exist == 1 {
+		return path, nil
+	}
+	if exist > 1 {
+		return path, fmt.Errorf("the resource name %s is not unique", resource)
+	}
+	return path, fmt.Errorf("there is no resource %s", resource)
+}
+
+func convert(sources map[string]resourceMetadata) map[string]string {
+	list := make(map[string]string)
+	for k, v := range sources {
+		if len(v.subResources) > 0 {
+			for sub, subWithPathName := range convert(v.subResources) {
+				if subWithPathName != "" {
+					withPathname := fmt.Sprintf("/%s:*%s", k, subWithPathName)
+					if v.pathName != "" {
+						withPathname = fmt.Sprintf("/%s:{%s}%s", k, v.pathName, subWithPathName)
+					}
+					list[fmt.Sprintf("/%s%s", k, sub)] = withPathname
+				}
+			}
+		}
+		withPathname := fmt.Sprintf("/%s:*/", k)
+		if v.pathName != "" {
+			withPathname = fmt.Sprintf("/%s:{%s}/", k, v.pathName)
+		}
+		list[fmt.Sprintf("/%s/", k)] = withPathname
+	}
+	return list
+}
+
+// registerResourceAction register resource actions
+func registerResourceAction(resource string, actions ...string) {
+	lock.Lock()
+	defer lock.Unlock()
+	if resourceActions == nil {
+		resourceActions = make(map[string][]string)
+	}
+	if _, exist := ResourceMaps[resource]; !exist {
+		path, err := checkResourcePath(resource)
+		if err != nil {
+			panic(fmt.Sprintf("resource %s is not exist", resource))
```

Comment:
Should we use panic here?

Reply:
This error represents a coding exception.